[wasm] Build improvements based on feedback

radical opened the pull request:
- Fix for a incremental build issue
- Change native build defaults for `Debug` config, to make it faster (thanks @SteveSandersonMS, @mattleibow)

- MonoAOTCompiler:
  - Skip unmanaged assemblies, and emit a warning
- PInvokeTableGenerator:
  - Fix to not fail in presence of native variadic functions, and pinvokes with function pointers (eg. used by sqlite)

radical committed:
Refactor to allow fast-path

radical committed:
MonoAOTCompiler: check for nothing-changed case

The task needs to run `mono-aot-cross` on all the assemblies, even if a
given assembly is older than its output files, so as to correctly handle
changes in its dependencies.

This commit tries to check if all the input assemblies are up-to-date with their
corresponding output files, and avoids doing any work if that's true.

radical committed:
[wasm] Change optimization flag defaults for Debug config

`WasmNativeStrip` -> `false`
Compile, and link optimization flags for `emcc` to `-O1`

radical committed:
[wasm] EmccCompile: Fix incremental build, in case of only partial

.. rebuilds. Specifically, when a rebuild causes only the *linking* part
to be run again. In that case, we were correctly skipping over compiling
native files, but didn't add them to `@(FileWrites)`, which caused
msbuild's incremental clean logic to treat them as "orphaned" files,
and delete them!

radical committed:
MonoAOTCompiler: Skip unmanaged assemblies, and emit a warning

radical committed:
Bump sdk for workload testing to 6.0.100-rc.2.21474.31

.. and remove the Blazor sdk targets hack used for testing, since the
updated SDK has the fixes.

radical committed:
MonoAOTCompiler: write the cache even when some files fail to compile

- Ensure that the cache gets written even when some of the files failed
  to compile
- And when some were skipped, and others failed

lewing commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -156,6 +157,7 @@
       <_EmccAssertionLevelDefault>0</_EmccAssertionLevelDefault>
       <_EmccOptimizationFlagDefault Condition="'$(_WasmDevel)' == 'true'">-O0 -s ASSERTIONS=$(_EmccAssertionLevelDefault)</_EmccOptimizationFlagDefault>
+      <_EmccOptimizationFlagDefault Condition="'$(_EmccOptimizationFlagDefault)' == '' and '$(Configuration)' == 'Debug'">-O1</_EmccOptimizationFlagDefault>
       <_EmccOptimizationFlagDefault Condition="'$(_EmccOptimizationFlagDefault)' == '' and '$(OS)' != 'Windows_NT' and '$(Configuration)' == 'Debug'">-Os</_EmccOptimizationFlagDefault>
```
The next line looks suspect too.

radical commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -156,6 +157,7 @@
       <_EmccOptimizationFlagDefault Condition="'$(_WasmDevel)' == 'true'">-O0 -s ASSERTIONS=$(_EmccAssertionLevelDefault)</_EmccOptimizationFlagDefault>
+      <_EmccOptimizationFlagDefault Condition="'$(_EmccOptimizationFlagDefault)' == '' and '$(Configuration)' == 'Debug'">-O1</_EmccOptimizationFlagDefault>
       <_EmccOptimizationFlagDefault Condition="'$(_EmccOptimizationFlagDefault)' == '' and '$(OS)' != 'Windows_NT' and '$(Configuration)' == 'Debug'">-Os</_EmccOptimizationFlagDefault>
       <_EmccOptimizationFlagDefault Condition="'$(_EmccOptimizationFlagDefault)' == '' and '$(Configuration)' != 'Debug'">-Oz</_EmccOptimizationFlagDefault>
```
```suggestion
```

radical committed:
Apply suggestions from code review

Co-authored-by: Larry Ewing <lewing@microsoft.com>

radical committed:
Don't set optimization defaults for Debug config, when publishing

radical committed:
Wasm.Build.Tests: Disable net5.0 because they can't be tested right now

https://github.com/dotnet/runtime/issues/59538

radical committed:
[wasm] Error for undefined symbols only when publishing

For non-publish builds, undefined symbols will show up as warnings:

```
EXEC : warning : undefined symbol: sqlite3_column_database_name (referenced by top-level compiled C/C++ code) [/BlazorWebAssemblySqlite.csproj]
EXEC : warning : undefined symbol: sqlite3_column_origin_name (referenced by top-level compiled C/C++ code) [/BlazorWebAssemblySqlite.csproj]
EXEC : warning : undefined symbol: sqlite3_column_table_name (referenced by top-level compiled C/C++ code) [/BlazorWebAssemblySqlite.csproj]
EXEC : warning : undefined symbol: sqlite3_snapshot_cmp (referenced by top-level compiled C/C++ code) [/BlazorWebAssemblySqlite.csproj]
EXEC : warning : undefined symbol: sqlite3_snapshot_free (referenced by top-level compiled C/C++ code) [/BlazorWebAssemblySqlite.csproj]
EXEC : warning : undefined symbol: sqlite3_snapshot_get (referenced by top-level compiled C/C++ code) [/BlazorWebAssemblySqlite.csproj]
EXEC : warning : undefined symbol: sqlite3_snapshot_open (referenced by top-level compiled C/C++ code) [/BlazorWebAssemblySqlite.csproj]
EXEC : warning : undefined symbol: sqlite3_snapshot_recover (referenced by top-level compiled C/C++ code) [/BlazorWebAssemblySqlite.csproj]
```

radical committed:
[wasm] PInvokeTableGenerator: Add support for variadic functions

Currently, for a variadic function like:

`int sqlite3_config(int, ...);`

.. and multiple pinvokes like:

```csharp
[DllImport(SQLITE_DLL, ExactSpelling=true, EntryPoint = "sqlite3_config", CallingConvention = CALLING_CONVENTION)]
public static extern unsafe int sqlite3_config_none(int op);

[DllImport(SQLITE_DLL, ExactSpelling=true, EntryPoint = "sqlite3_config", CallingConvention = CALLING_CONVENTION)]
public static extern unsafe int sqlite3_config_int(int op, int val);

[DllImport(SQLITE_DLL, ExactSpelling=true, EntryPoint = "sqlite3_config", CallingConvention = CALLING_CONVENTION)]
public static extern unsafe int sqlite3_config_log(int op, NativeMethods.callback_log func, hook_handle pvUser);
```

.. we generate:

```c
int sqlite3_config (int);
int sqlite3_config (int,int);
int sqlite3_config (int,int,int);
```

.. which fails to compile.

Instead, this patch will generate a variadic declaration with one fixed
parameter:

```c
// Variadic signature created for
//   System.Int32 sqlite3_config_none(System.Int32)
//   System.Int32 sqlite3_config_int(System.Int32, System.Int32)
//   System.Int32 sqlite3_config_log(System.Int32, SQLitePCL.SQLite3Provider_e_sqlite3+NativeMethods+callback_log, SQLitePCL.hook_handle)
int sqlite3_config (int, ...);
```

TODO: functions with different first argument

radical committed:
[wasm] Handle pinvokes with function pointers

- For pinvokes with function pointers, *no* declaration is added to
  `pinvoke-table.h`, and a warning is raised:

```
warning : DllImports with function pointers are not supported. Calling them will fail. Managed DllImports:  [/Users/radical/dev/r2/artifacts/bin/Wasm.Build.Tests/net6.0-Release/browser-wasm/g3acrk4b.a0o/variadic_g3acrk4b.a0o.csproj]
warning :    Type: Test, Method: System.Int32 using_sum_one(?) [/Users/radical/dev/r2/artifacts/bin/Wasm.Build.Tests/net6.0-Release/browser-wasm/g3acrk4b.a0o/variadic_g3acrk4b.a0o.csproj]
```

- Also, handle multiple pinvokes with the same number of params, but
  different types

radical committed:
[wasm] PInvokeTableGenerator: handle pinvokes with function pointers

- pinvokes with function pointers could be for variadic functions too
- if there are multiple pinvokes for the same native function, but some
  of them have function pointers, then ignore those but generate the
  declaration for the rest

- raise better warnings
- and emit info about the variadic pinvokes in pinvoke-table.h

radical committed:
add missing variadic.{c,o}

radical committed:
[wasm] Add test for issue dotnet#59255

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-improvements

radical committed:
MonoAOTCompiler: Check the hash for the file also, for "all up-to-date" check

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-improvements

radical committed:
Revert "MonoAOTCompiler: Check the hash for the file also, for "all up-to-date" check"

Newer timestamp, but unchanged file is an indication of a build issue,
and doesn't need to be handled in this task.

This reverts commit 712dbb3280ed84d212410c77a43b4be9593fb372.

radical committed:
PInvokeTableGenerator: don't generate any decl for variadic functions

Based on feedback from Zoltan

radical commented:
wasm test failure is https://github.com/dotnet/runtime/issues/48717

radical commented:
Coreclr failures are https://github.com/dotnet/runtime/issues/10680 , unrelated to this PR.

radical committed:
Add missing `using` for disposable objects.

radical committed:
cleanup

radical committed:
Merge remote-tracking branch 'origin/main' into rf-wasm-improvements

radical committed:
address feedback from @lewing

radical commented:
Android failure is https://github.com/dotnet/runtime/issues/59779 .

radical commented:
/backport release/6.0

radical commented:
/backport to release/6.0
